File: Module 1/main.cpp
#include <chrono>
#include <csignal>
#include <exception>
#include <iomanip>
#include <iostream>
#include <optional>
#include <string>
#include <thread>
#include <vector>

#include "ConfigLoader.hpp"
#include "Renderer.hpp"
#include "Scenarios.hpp"
#include "Simulation.hpp"

struct CliOptions {
    std::optional<std::string> scenario;
    std::optional<std::string> file;
    int steps = 2000;
    std::optional<double> dt;
    int frameEvery = 10;
    double delay = 0.05;
    bool noRender = false;
};

bool g_interrupted = false;

void handleSigInt(int) {
    g_interrupted = true;
}

void printHelp() {
    std::cout
        << "N-body gravity simulator (C++)\n\n"
        << "Usage:\n"
        << "  nbody                              # interactive menu\n"
        << "  nbody --scenario solar_system\n"
        << "  nbody --file example_scenario.json\n"
        << "  nbody --steps 500 --dt 7200 --no-render\n\n"
        << "Options:\n"
        << "  --scenario <name>    Preset: solar_system | binary_stars | figure_eight | random_cloud\n"
        << "  --file <path>        Load scenario from JSON config file\n"
        << "  --steps <n>          Number of simulation steps (default: 2000)\n"
        << "  --dt <seconds>       Override time-step\n"
        << "  --frame-every <n>    Render every n steps (default: 10)\n"
        << "  --delay <seconds>    Delay between frames (default: 0.05)\n"
        << "  --no-render          Headless mode\n"
        << "  --help               Show this help\n";
}

CliOptions parseArgs(int argc, char** argv) {
    CliOptions opts;

    for (int i = 1; i < argc; ++i) {
        const std::string arg = argv[i];
        auto needValue = [&](const std::string& name) -> std::string {
            if (i + 1 >= argc) {
                throw std::runtime_error("Missing value for " + name);
            }
            return std::string(argv[++i]);
        };

        if (arg == "--help") {
            printHelp();
            std::exit(0);
        } else if (arg == "--scenario" || arg == "-s") {
            opts.scenario = needValue(arg);
        } else if (arg == "--file" || arg == "-f") {
            opts.file = needValue(arg);
        } else if (arg == "--steps" || arg == "-n") {
            opts.steps = std::stoi(needValue(arg));
        } else if (arg == "--dt") {
            opts.dt = std::stod(needValue(arg));
        } else if (arg == "--frame-every") {
            opts.frameEvery = std::stoi(needValue(arg));
        } else if (arg == "--delay") {
            opts.delay = std::stod(needValue(arg));
        } else if (arg == "--no-render") {
            opts.noRender = true;
        } else {
            throw std::runtime_error("Unknown argument: " + arg);
        }
    }

    return opts;
}

std::string interactiveScenarioChoice() {
    const auto& reg = Scenarios::registry();
    std::vector<std::string> keys;
    keys.reserve(reg.size());

    std::cout << "\nN-Body Gravity Simulator (C++)\n\n";
    std::cout << "Choose a scenario:\n";

    int idx = 1;
    for (const auto& entry : reg) {
        keys.push_back(entry.first);
        std::cout << "  " << idx++ << ". " << entry.second.first << "\n";
    }
    std::cout << "  " << idx << ". Load from JSON file\n\n";

    while (true) {
        std::cout << "Enter choice [1-" << idx << "]: ";
        int choice = 0;
        if (!(std::cin >> choice)) {
            std::cin.clear();
            std::cin.ignore(10000, '\n');
            std::cout << "Invalid choice.\n";
            continue;
        }

        if (choice >= 1 && choice <= static_cast<int>(keys.size())) {
            return keys[static_cast<std::size_t>(choice - 1)];
        }
        if (choice == idx) {
            return "__file__";
        }
        std::cout << "Invalid choice.\n";
    }
}

void printEnergyReport(const Simulation& sim, double initialEnergy) {
    const double ke = sim.totalKineticEnergy();
    const double pe = sim.totalPotentialEnergy();
    const double total = ke + pe;
    const double drift = (initialEnergy != 0.0)
        ? std::abs((total - initialEnergy) / initialEnergy) * 100.0
        : 0.0;

    std::cout << "\n--------------------------------------------------\n";
    std::cout << std::scientific << std::setprecision(6);
    std::cout << "  Kinetic Energy   : " << ke << " J\n";
    std::cout << "  Potential Energy : " << pe << " J\n";
    std::cout << "  Total Energy     : " << total << " J\n";
    std::cout << "  Initial Energy   : " << initialEnergy << " J\n";
    std::cout << std::fixed << std::setprecision(4);
    std::cout << "  Energy drift     : " << drift << " %\n";

    const auto& log = sim.getCollisionLog();
    if (!log.empty()) {
        std::cout << "\n  Collision events (" << log.size() << "):\n";
        for (const auto& entry : log) {
            std::cout << "    " << entry << "\n";
        }
    }
    std::cout << "--------------------------------------------------\n";
}

void runSimulation(Simulation& sim, int steps, bool render, int frameEvery, double delaySeconds) {
    Renderer renderer;

    sim.computeAccelerations();
    const double initialEnergy = sim.totalEnergy();

    std::cout << "\nStarting simulation: " << sim.getBodies().size()
              << " bodies, dt=" << sim.getDt() << "s, steps=" << steps << "\n";
    if (!render) {
        std::cout << "(headless mode)\n";
    }

    for (int step = 0; step < steps && !g_interrupted; ++step) {
        sim.step();

        if (render && (step % frameEvery == 0)) {
            renderer.render(sim.getBodies(), sim.getTimeElapsed(), sim.totalEnergy(), sim.getCollisionLog());
            std::this_thread::sleep_for(std::chrono::duration<double>(delaySeconds));
        }

        if (sim.getBodies().empty()) {
            std::cout << "All bodies merged.\n";
            break;
        }
    }

    if (g_interrupted) {
        std::cout << "\nInterrupted by user.\n";
    }

    printEnergyReport(sim, initialEnergy);
}

int main(int argc, char** argv) {
    std::signal(SIGINT, handleSigInt);

    try {
        const CliOptions opts = parseArgs(argc, argv);

        Simulation sim;

        if (opts.file.has_value()) {
            sim = ConfigLoader::loadScenario(*opts.file);
        } else {
            std::string scenarioName;

            if (opts.scenario.has_value()) {
                scenarioName = *opts.scenario;
            } else {
                const std::string choice = interactiveScenarioChoice();
                if (choice == "__file__") {
                    std::cout << "Path to JSON file: ";
                    std::string path;
                    std::cin >> path;
                    sim = ConfigLoader::loadScenario(path);
                } else {
                    scenarioName = choice;
                }
            }

            if (!scenarioName.empty()) {
                const auto& reg = Scenarios::registry();
                const auto it = reg.find(scenarioName);
                if (it == reg.end()) {
                    throw std::runtime_error("Unknown scenario: " + scenarioName);
                }

                double dt = opts.dt.value_or(3600.0);
                if (scenarioName == "figure_eight" && !opts.dt.has_value()) {
                    dt = 10.0;
                }
                sim = it->second.second(dt);
            }
        }

        if (opts.dt.has_value()) {
            sim.setDt(*opts.dt);
        }

        runSimulation(sim, opts.steps, !opts.noRender, opts.frameEvery, opts.delay);
        return 0;
    } catch (const std::exception& ex) {
        std::cerr << "Error: " << ex.what() << "\n";
        return 1;
    }
}

File: Module 1/main.py
"""
main.py — N-body gravity simulator  (Days 1-14 complete)

Usage
-----
    python main.py                          # interactive menu
    python main.py --scenario solar_system  # run a preset directly
    python main.py --file my_scenario.json  # load a JSON config
    python main.py --steps 500 --dt 7200    # override step count / time-step
    python main.py --no-render              # headless run (energy check only)

Controls while running
----------------------
    Press Ctrl-C to stop at any time.
"""

import argparse
import sys
import time

from simulation import Simulation
from renderer import Renderer
from scenarios import SCENARIOS
from config_loader import load_scenario


# ---------------------------------------------------------------------------
# Helpers
# ---------------------------------------------------------------------------

def _print_menu() -> str:
    """Print the interactive scenario menu and return the user's choice."""
    print("\n╔══════════════════════════════════════╗")
    print("║      N-Body Gravity Simulator        ║")
    print("╚══════════════════════════════════════╝\n")
    print("Select a scenario:\n")
    keys = list(SCENARIOS.keys())
    for i, key in enumerate(keys, 1):
        description, _ = SCENARIOS[key]
        print(f"  {i}. {description}")
    print(f"  {len(keys)+1}. Load from JSON file")
    print()
    while True:
        raw = input(f"Enter choice [1-{len(keys)+1}]: ").strip()
        if raw.isdigit():
            choice = int(raw)
            if 1 <= choice <= len(keys) + 1:
                if choice <= len(keys):
                    return keys[choice - 1]
                return "__file__"
        print("  Invalid choice, try again.")


def _energy_report(sim: Simulation, initial_energy: float) -> None:
    """Print an energy conservation summary."""
    ke = sim.total_kinetic_energy()
    pe = sim.total_potential_energy()
    total = ke + pe
    drift = abs((total - initial_energy) / initial_energy) * 100 if initial_energy != 0 else 0.0
    print(f"\n{'─'*50}")
    print(f"  Kinetic Energy   : {ke:.6e} J")
    print(f"  Potential Energy : {pe:.6e} J")
    print(f"  Total Energy     : {total:.6e} J")
    print(f"  Initial Energy   : {initial_energy:.6e} J")
    print(f"  Energy drift     : {drift:.4f} %")
    if sim.collision_log:
        print(f"\n  Collision events ({len(sim.collision_log)}):")
        for entry in sim.collision_log:
            print(f"    {entry}")
    print(f"{'─'*50}")


# ---------------------------------------------------------------------------
# Main runner
# ---------------------------------------------------------------------------

def run(
    sim: Simulation,
    steps: int,
    render: bool,
    frame_every: int,
    frame_delay: float,
) -> None:
    """Run the simulation loop with optional ASCII rendering."""
    renderer = Renderer() if render else None

    # Prime initial accelerations and record starting energy
    from simulation import G
    sim._compute_accelerations()
    initial_energy = sim.total_energy()

    print(f"\nStarting simulation: {len(sim.bodies)} bodies, "
          f"dt={sim.dt}s, {steps} steps")
    if not render:
        print("(headless mode — no display)")

    try:
        for step in range(steps):
            sim.step()
            if render and (step % frame_every == 0):
                energy = sim.total_energy()
                renderer.render(
                    sim.bodies,
                    sim.time_elapsed,
                    energy=energy,
                    collision_log=sim.collision_log,
                )
                time.sleep(frame_delay)
                if not sim.bodies:
                    print("All bodies merged — simulation complete.")
                    break
    except KeyboardInterrupt:
        print("\n[Interrupted by user]")

    _energy_report(sim, initial_energy)


# ---------------------------------------------------------------------------
# CLI entry point
# ---------------------------------------------------------------------------

def main() -> None:
    parser = argparse.ArgumentParser(
        description="N-body gravity simulator",
        formatter_class=argparse.RawDescriptionHelpFormatter,
        epilog=__doc__,
    )
    parser.add_argument(
        "--scenario", "-s",
        choices=list(SCENARIOS.keys()),
        help="Preset scenario to run.",
    )
    parser.add_argument(
        "--file", "-f",
        metavar="PATH",
        help="Path to a JSON scenario file.",
    )
    parser.add_argument(
        "--steps", "-n",
        type=int,
        default=2000,
        help="Number of integration steps (default: 2000).",
    )
    parser.add_argument(
        "--dt",
        type=float,
        default=None,
        help="Override the time-step in seconds.",
    )
    parser.add_argument(
        "--frame-every",
        type=int,
        default=10,
        help="Render every N steps (default: 10).",
    )
    parser.add_argument(
        "--delay",
        type=float,
        default=0.05,
        help="Seconds to pause between rendered frames (default: 0.05).",
    )
    parser.add_argument(
        "--no-render",
        action="store_true",
        help="Disable ASCII rendering (headless energy-check mode).",
    )

    args = parser.parse_args()

    # ---- Build simulation ------------------------------------------------
    sim: Simulation | None = None

    if args.file:
        try:
            sim = load_scenario(args.file)
        except (FileNotFoundError, ValueError) as exc:
            print(f"Error loading '{args.file}': {exc}", file=sys.stderr)
            sys.exit(1)

    elif args.scenario:
        _, factory = SCENARIOS[args.scenario]
        dt = args.dt if args.dt else 3600.0
        # figure_eight needs a much smaller dt
        if args.scenario == "figure_eight":
            dt = args.dt if args.dt else 10.0
        sim = factory(dt=dt)

    else:
        # Interactive menu
        choice = _print_menu()
        if choice == "__file__":
            path = input("Enter path to JSON file: ").strip()
            try:
                sim = load_scenario(path)
            except (FileNotFoundError, ValueError) as exc:
                print(f"Error: {exc}", file=sys.stderr)
                sys.exit(1)
        else:
            _, factory = SCENARIOS[choice]
            dt = args.dt if args.dt else (10.0 if choice == "figure_eight" else 3600.0)
            sim = factory(dt=dt)

    if args.dt and sim:
        sim.dt = args.dt

    # ---- Run ---------------------------------------------------------------
    run(
        sim=sim,
        steps=args.steps,
        render=not args.no_render,
        frame_every=args.frame_every,
        frame_delay=args.delay,
    )


if __name__ == "__main__":
    main()
